src/library.cpp

```diff
@@ -158,6 +158,83 @@ bool Library::removeBookmark(const std::string& zimId, const std::string& url)
 }
 
 
+int Library::resolveBookmarks(const std::string& source, std::string destination) {
+  // We don't have a source. Let's loop on all books (and not bookmarks) which have invalid bookmarks.
+  if (source.empty()) {
+    std::set<std::string> sourceBooks;
+    {
+      std::lock_guard<std::recursive_mutex> lock(m_mutex);
+      for(auto& bookmark:m_bookmarks) {
+        if (m_books.find(bookmark.getBookId()) == m_books.end()) {
+          sourceBooks.insert(bookmark.getBookId());
+        }
+      }
+    }
+    int changed = 0;
+    for(auto& sourceBook:sourceBooks) {
+      changed += resolveBookmarks(sourceBook);
+    }
+    return changed;
+  }
+
+  // We have a source book. Change all its bookmarks (even if the book itself is in the library).
+  std::lock_guard<std::recursive_mutex> lock(m_mutex);
+
+  Bookmark firstBookmarkToChange;
+  for(auto& bookmark:m_bookmarks) {
+    if (bookmark.getBookId() == source) {
+      firstBookmarkToChange = bookmark;
+      break;
+    }
+  }
+
+  if (firstBookmarkToChange.getBookId().empty()) {
+    return 0;
+  }
+
+  if (destination.empty()) {
+    // Search for a existing book with the same name
+    auto book_filter = Filter();
+    if (!firstBookmarkToChange.getBookName().empty()) {
+      book_filter.name(firstBookmarkToChange.getBookName());
+    } else {
+      // We don't have a name stored (older bookmarks)
+      // Fallback on title (All bookmarks should have one, but let's be safe against wrongly filled bookmark)
+      if (firstBookmarkToChange.getBookTitle().empty()) {
+          // No bookName nor bookTitle, no way to find target book.
+          return 0;
+      }
+      book_filter.query("title:\""+firstBookmarkToChange.getBookTitle() + "\"");
+    }
+    auto targetBooks = filter(book_filter);
+    // Remove source book
+    auto foundIT = std::find(targetBooks.begin(), targetBooks.end(), source);
+    if (foundIT != targetBooks.end()) {
+      targetBooks.erase(foundIT);
+    }
+    
+    if (targetBooks.empty()) {
+      // No existing book found for the target, or the same than source.
+      return 0;
+    }
+    if (targetBooks.size() != 1) {
+      sort(targetBooks, DATE, false);
+    }
+    destination = targetBooks[0];
```

Comment:
This piece of code deserves being packaged into a (private) named function `std::string Library::getMostRecentTargetBookId(const Bookmark&) const` (the name could be different but it must correctly reflect the semantics of that function).